src/Functions/minSampleSize.cpp

```diff
@@ -0,0 +1,166 @@
+#include <Columns/ColumnTuple.h>
+#include <Columns/ColumnsNumber.h>
+#include <DataTypes/DataTypesNumber.h>
+#include <DataTypes/DataTypeTuple.h>
+#include <Functions/IFunction.h>
+#include <Functions/FunctionFactory.h>
+#include <Functions/FunctionHelpers.h>
+#include <boost/math/distributions/normal.hpp>
+#include <cmath>
+#include <cfloat>
+
+
+namespace DB
+{
+
+    template <typename Impl>
+    class FunctionMinSampleSize : public IFunction
+    {
+    public:
+        static constexpr auto name = Impl::name;
+
+        static FunctionPtr create(ContextPtr)
+        {
+            return std::make_shared<FunctionMinSampleSize<Impl>>();
+        }
+
+        String getName() const override
+        {
+            return name;
+        }
+
+        size_t getNumberOfArguments() const override { return Impl::num_args; }
+
+        bool useDefaultImplementationForNulls() const override { return false; }
+        bool useDefaultImplementationForConstants() const override { return true; }
+        bool isSuitableForShortCircuitArgumentsExecution(const DataTypesWithConstInfo & /*arguments*/) const override { return false; }
+
+        static DataTypePtr getReturnType()
+        {
+            DataTypes types
+            {
+                std::make_shared<DataTypeNumber<UInt64>>(),
+                std::make_shared<DataTypeNumber<Float64>>(),
+                std::make_shared<DataTypeNumber<Float64>>(),
+            };
+
+            Strings names
+            {
+                "minimum_sample_size",
+                "detect_range_lower",
+                "detect_range_upper",
+            };
+
+            return std::make_shared<DataTypeTuple>(
+                std::move(types),
+                std::move(names)
+            );
+        }
+
+        DataTypePtr getReturnTypeImpl(const DataTypes & /*arguments*/) const override
+        {
+            return getReturnType();
+        }
+
+        ColumnPtr executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr &, size_t input_rows_count) const override
+        {
+            MutableColumnPtr to{getReturnType()->createColumn()};
```

Comment:
Current implementation is extremely slow.
First we can safely unwrap arguments to columns implementations, ColumnFloatUInt64, ColumnFloat64,  ColumnFloat64.
Second we can unwrap result tuple column in 3 columns UInt64, Float64, Float64, resize data to necessary rows count, and then just insert by index.

Reply:
I unwrapped arguments and the result tuple.